README.md
## Sensu-Plugins-influxdb

[![Build Status](https://travis-ci.org/sensu-plugins/sensu-plugins-influxdb.svg?branch=master)](https://travis-ci.org/sensu-plugins/sensu-plugins-influxdb)
[![Gem Version](https://badge.fury.io/rb/sensu-plugins-influxdb.svg)](http://badge.fury.io/rb/sensu-plugins-influxdb)
[![Code Climate](https://codeclimate.com/github/sensu-plugins/sensu-plugins-influxdb/badges/gpa.svg)](https://codeclimate.com/github/sensu-plugins/sensu-plugins-influxdb)
[![Test Coverage](https://codeclimate.com/github/sensu-plugins/sensu-plugins-influxdb/badges/coverage.svg)](https://codeclimate.com/github/sensu-plugins/sensu-plugins-influxdb)
[![Dependency Status](https://gemnasium.com/sensu-plugins/sensu-plugins-influxdb.svg)](https://gemnasium.com/sensu-plugins/sensu-plugins-influxdb)
[![Sensu Bonsai Asset](https://img.shields.io/badge/Bonsai-Download%20Me-brightgreen.svg?colorB=89C967&logo=sensu)](https://bonsai.sensu.io/assets/sensu-plugins/sensu-plugins-influxdb)

## Sensu Asset  
  The Sensu assets packaged from this repository are built against the Sensu ruby runtime environment. When using these assets as part of a Sensu Go resource (check, mutator or handler), make sure you include the corresponding Sensu ruby runtime asset in the list of assets needed by the resource.  The current ruby-runtime assets can be found [here](https://bonsai.sensu.io/assets/sensu/sensu-ruby-runtime) in the [Bonsai Asset Index](bonsai.sensu.io).

## Functionality

## Files
 * bin/check-influxdb.rb
 * bin/check-influxdb-query.rb
 * bin/metrics-influxdb.rb
 * bin/mutator-influxdb-line-protocol.rb 

## Usage - metrics-influxdb.rb
Add the following to `/etc/sensu/conf.d/influx.json` "plugin expects `influxdb` to be a top level node in the json
**metrics-influxdb**
```
{
    "influxdb": {
        "hosts"         : ["influxdb.familyguy.com"],
        "host"          : "influxdb.familyguy.com",
        "port"          : "8086",
        "username"      : "root",
        "password"      : "root",
        "database"      : "stats",
        # OPTIONAL
        "time_precision": "s",
        "use_ssl"       : false,
        "verify_ssl"    : true,
        "ssl_ca_cert"   : "path to the ca certificate file",
        "auth_method"   : "params",
        "initial_delay" : 0.01,
        "max_delay"     : 30,
        "open_timeout"  : 5,
        "read_timeout"  : 300,
        "retry"         : null,
        "prefix"        : "",
        "denormalize"   : true,
        "status"        : true
    }
}
```
Then add the following to your `/etc/sensu/conf.d/handlers.json`:
```
{
  "handlers": {
     "influx-tcp": {
       "type": "pipe",
       "command": "/opt/sensu/embedded/bin/metrics-influxdb.rb"
     }
   }
}
```

Setting "status" to true will store the metric using the status as the value and the check name as the series

## Usage - mutator-influxdb-line-protocol.rb
Add the mutator-influxdb-line-protocol.rb file into /etc/sensu/extensions/

Then add the following to your `/etc/sensu/conf.d/handlers.json`:
```
{
  "handlers": {
    "influxdb_udp": {
      "type": "udp",
      "socket": {
        "host": "localhost",
        "port": 8090
      },
      "mutator": "influxdb_line_protocol"
    }
  }
}
```

## Check configuration
To ship additional tags to your metrics via mutator-influxdb-line-protocol.rb (only available in influxdb >= 0.9), just add a tags block inside your check.

```
{
  "checks": {
    "cpu-metrics": {
      "command": "/opt/sensu/embedded/bin/metrics-cpu-pcnt-usage.rb",
      "handlers": [
        "influxdb_udp"
      ],
      "interval": 60,
      "subscribers": [
        "base"
      ],
      "type": "metric",
      "standalone": false,
      "tags": {
        "group": "operations"
      }
    }
  }
}
```

## Installation

[Installation and Setup](http://sensu-plugins.io/docs/installation_instructions.html)

## Notes


CHANGELOG.md
# Change Log
This project adheres to [Semantic Versioning](http://semver.org/).

This CHANGELOG follows the format listed [here](https://github.com/sensu-plugins/community/blob/master/HOW_WE_CHANGELOG.md)

## [Unreleased]
### Breaking Changes
- Return ok for empty result irrespective of jsonpath expression

### Added
- Updated asset build targets to support centos6
- Removed centos from bonsai asset definition
- Fix allow 'any' mode

## [2.0.0] - 2019-04-18
### Breaking Changes
- Update minimum required ruby version to 2.3. Drop unsupported ruby versions.
- Bump `sensu-plugin` dependency from `~> 1.3` to `~> 4.0` you can read the changelog entries for [4.0](https://github.com/sensu-plugins/sensu-plugin/blob/master/CHANGELOG.md#400---2018-02-17), [3.0](https://github.com/sensu-plugins/sensu-plugin/blob/master/CHANGELOG.md#300---2018-12-04), and [2.0](https://github.com/sensu-plugins/sensu-plugin/blob/master/CHANGELOG.md#v200---2017-03-29)

### Added
- Travis build automation to generate Sensu Asset tarballs that can be used n conjunction with Sensu provided ruby runtime assets and the Bonsai Asset Index
- Require latest sensu-plugin for [Sensu Go support](https://github.com/sensu-plugins/sensu-plugin#sensu-go-enablement)

## [1.4.0] - 2019-02-19
### Added
- Added "sum" mode for check-influxdb.query check

## [1.3.0] - 2018-08-06
### Fixed
- when hostname contains a `.` such as having a FQDN we rely on taking the remaining string rather than doing a split (@cholletk)
- misc typo in PR template (@majormoses)

### Added
- Ruby 2.4.1 testing

### Changed
- updated changelog guidlines location (@majormoses)

## [1.2.0] - 2017-05-15
- Added "mode" into check-influxdb-query check to allow multiple values in a query

## [1.1.0] - 2016-12-01
### Added
- Added tags support for mutator-influxdb-line-protocol.rb
- Added retry support for check-influxdb-query.rb - The InfluxDB gem by default retries indefinitely and will cause the query to hang.  A retry of 12 will retry for approx 37 seconds.

### Changed
- Update to `influxdb` gem 0.3.13

## [1.0.0] - 2016-10-06
### Added
- Ruby 2.3.0 support

### Removed
- Ruby 1.9.3 support

### Changed
- Update `influxdb` dep to 0.3.10
- Update `dentaku` dep to 2.0.9
- Update `sensu-plugin` dep to ~> 1.3
- Update `jsonpath` dep to 0.5.8
- Update to rubocop 0.40 and cleanup

### [0.0.7] - 2016-04-26
- Line protocol mutator extension
- added tags support for metrics-influxdb.rb
- Fixed option tag in check-influxdb-query
- added support for storing check status for metrics-influxdb.rb
- metrics-influxdb.rb will now create the database if needed
- metrics-influxdb.rb can now be configured to use a specific settings block

### [0.0.6] - 2016-04-18
- supress warning if only checking for an existing return value in check-influxdb-query.

### [0.0.5] - 2015-10-19
- added support for https in check-influxdb.
- pass ssl arguments to the influxdb object in check-influxdb-query.
- renamed the config option in check-influxdb from ssl to verify_ssl.
- changed README to update the config option ssl_ca_cert.

### [0.0.4] - 2015-08-10
- updated metrics-influxdb.rb to symbolize keys for InfluxDB::Client
- updated influxdb gem to 0.2.2
- updated metrics-influxdb.rb to support influxdb 0.9.x
- updated check-influxdb.rb and check-influxdb-query.rb to support influxdb 0.9.x
- updated README.md file

## [0.0.3] - 2015-07-14
### Changed
- updated sensu-plugin gem to 1.2.0

## [0.0.2] - 2015-06-03
### Fixed
- added binstubs

### Changed
- removed cruft from /lib

## 0.0.1 - 2015-04-29
### Added
- initial release

[Unreleased]: https://github.com/sensu-plugins/sensu-plugins-influxdb/compare/2.0.0...HEAD
[2.0.0]: https://github.com/sensu-plugins/sensu-plugins-influxdb/compare/1.4.0...2.0.0
[1.4.0]: https://github.com/sensu-plugins/sensu-plugins-influxdb/compare/1.3.0...1.4.0
[1.3.0]: https://github.com/sensu-plugins/sensu-plugins-influxdb/compare/1.2.0...1.3.0
[1.2.0]: https://github.com/sensu-plugins/sensu-plugins-influxdb/compare/1.1.0...1.2.0
[1.1.0]: https://github.com/sensu-plugins/sensu-plugins-influxdb/compare/1.0.0...1.1.0
[1.0.0]: https://github.com/sensu-plugins/sensu-plugins-influxdb/compare/0.0.7...1.0.0
[0.0.7]: https://github.com/sensu-plugins/sensu-plugins-influxdb/compare/0.0.6...0.0.7
[0.0.6]: https://github.com/sensu-plugins/sensu-plugins-influxdb/compare/0.0.5...0.0.6
[0.0.5]: https://github.com/sensu-plugins/sensu-plugins-influxdb/compare/0.0.4...0.0.5
[0.0.4]: https://github.com/sensu-plugins/sensu-plugins-influxdb/compare/0.0.3...0.0.4
[0.0.3]: https://github.com/sensu-plugins/sensu-plugins-influxdb/compare/0.0.2...0.0.3
[0.0.2]: https://github.com/sensu-plugins/sensu-plugins-influxdb/compare/0.0.1...0.0.2


scripts/code_deploy.rb
#! /usr/bin/env ruby

plugin = File.basename(File.expand_path('.'))
spec = Gem::Specification.load("#{ plugin }.gemspec")
lib = File.expand_path('../lib')
version_file = "lib/#{ plugin }/version.rb"

$LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)
require_relative "../../#{ plugin }/lib/#{ plugin }"
require 'date'
require 'json'
require 'base64'

## Environment Setup

# Drop the necessary keys into the build environment.
# Environment variables are not used due to the design of codeship, each project
# has its own set of variables so a key would need to be added or changed
# in ~160 repos and that just unpleasent to think about.
`key_deploy`

# the following chmod commands will go away when I can give Golang
# some more love
`chmod 0600 /home/rof/.gem/credentials`
`chmod 0600 /home/rof/.ssh/github_auto_key`
`chmod 0600 /home/rof/.ssh/gem-private_key.pem`
`chmod 0600 /home/rof/.ssh/git_token`

# This is needed for codeship as it checkouts a local branch, we want to
# ensure that we commit back up to master.
# The user.name maps to a Github machine user and the email is not necessary
`git checkout master`
`git fetch origin "+refs/heads/*:refs/remotes/origin/*"`
`git remote add repo git@github.com:sensu-plugins/#{ plugin }.git`
`git config --global user.email 'no-op@example.com'`
`git config --global user.name 'sensu-plugin'`

# set the git api token
github_token = File.read('/home/rof/.ssh/git_token')

#
# Build a gem and deploy it to rubygems
#
def deploy_rubygems(spec, plugin)
  `gem build #{ plugin }.gemspec`
  `gem push #{ spec.full_name }.gem`
end

#
# Create Github tag and release
#
def create_github_release(spec, plugin, github_token)
  `curl -H "Authorization: token #{ github_token }" -d '{ "tag_name": "#{ spec.version }", "target_commitish": "#{ ENV['CI_COMMIT_ID'] }", "name": "#{ spec.version }", "body": "#{ ENV['CI_MESSAGE'] }", "draft": "#{ spec.metadata['release_draft']}", "prerelease": "#{ spec.metadata['release_prerelease']}" }' https://api.github.com/repos/#{ github_org }/#{ plugin }/releases` # rubocop:disable all
end

#
# Bump the patch version of the plugin
#
def version_bump(version_file)
  # Read the file, bump the PATCH version
  contents = File.read(version_file).gsub(/(PATCH = )(\d+)/) { |_| Regexp.last_match[1] + (Regexp.last_match[2].to_i + 1).to_s }

  # Write the new contents of the file
  File.open(version_file, 'w') { |file| file.puts contents }
end

# create a github commit for the version bump
# the skip-ci flag is specific to codeship to prevent this from being run as a test
# Travis-CI will still run it as a test though
#
def create_github_commit(plugin)
  `git add lib/#{ plugin }/version.rb`
  `git commit -m 'version bump --skip-ci'`
  `git push repo master`
end

if ENV['CI_MESSAGE'] == 'deploy'
  version_bump(version_file)
  create_github_commit(plugin)
  deploy_rubygems(spec, plugin)
  create_github_release(spec, plugin, github_token)
end


extensions/influxdb_line_protocol.rb
# DESCRIPTION:
# Mutates check results to conform to InfluxDB's line protocol format
#
# Place this file in /etc/sensu/extensions and modify your handlers JSON config
#
# handlers.json
# {
#   "influxdb_udp": {
#      "type": "udp",
#      "mutator": "influxdb_line_protocol",
#      "socket": {
#        "host": "mgt-monitor-db1",
#        "port": 8090
#      }
#    }
# }
require "sensu/extension"

module Sensu
  module Extension
    class InfluxDBLineProtocol < Mutator
      def name
        "influxdb_line_protocol"
      end

      def description
        "returns check output formatted for InfluxDB's line protocol"
      end

      def run(event)
        host = event[:client][:name]
        metric = event[:check][:name]
        output = event[:check][:output]

        data = []
        output.split("\n").each do |result|
          m = result.split
          next unless m.count == 3
          key = m[0].split('.', 2)[1]
          key.gsub!('.', '_')
          value = m[1].to_f
          time = m[2].ljust(19, '0')
          data << "#{key},host=#{host},metric=#{metric} value=#{value} #{time}"
        end

        yield data.join("\n"), 0
      end
    end
  end
end
